### PR TITLE
salt-ssh: randomize name of state archive

### PR DESCRIPTION
**Pull request by mephi42**
### What does this PR do?
Fixes the issue when using more than one instance of salt-ssh at a
time resulted in 0 states being applied.

### What issues does this PR fix or reference?
#27898

### Previous Behavior
When two salt-ssh instances run concurrently, one of them ends with 0 states ran.

### New Behavior
Both instances complete successfully.

### Tests written?
No

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by cachedout** [comment]
I think this is fine but would also like @thatch45 to take a look.

**Review by thatch45 — CHANGES_REQUESTED** [review]
If I am not mistaken the cleanup routine could create another collision because it cleans up all state pkgs and not just the one it needs.
I would argue that we likely need to make the uuid up in the pkg_ssh function and send it down to pkg so that we have a consistent track of what the package name is that we are cleaning up. It might make sense to make the whole path higher up and pass it down so that we have less copy pasta here.

**Comment by mephi42** [comment]
@thatch45 Thanks for looking at this!

My reasoning was that we must make up the name on salt-ssh master, because otherwise collisions may occur on upload. Can this be avoided?

`_cleanup_ssh_pkgs` can indeed lead to problems when >10 concurrent salt-ssh instances are running. While unlikely in practice, I would prefer the design that would avoid this altogether. Since the ability to look at N state archives looks valuable, what do you think about the following scheme?

* Upload archive as `salt_state{1}.tgz`
* Pass it to `pkg_ssh` and `pkg` under this name
* When `pkg` exits, rename to `salt_state_debug{1}.tgz`
* Run cleanup on `salt_state_debug*.tgz` to keep the last N ones

I also agree that getting rid of copy-paste in `salt/client/ssh/wrapper/state.py` would be nice. I can do this once we decide how to proceed here.

**Comment by thatch45** [comment]
I think your solution sounds great because it allows existing apis to remain consistent without mandating that higher level interfaces implement the uuid creation. It also cleanly moves those into a location that can be cleaned effectively.

**Comment by mephi42** [comment]
@cachedout, yep, I pushed a new patch a few days ago and currently waiting for @thatch45 's comments.

**Comment by rallytime** [comment]
@thatch45 When you get a chance, can you take a look at the changes here?

**Review by thatch45 — APPROVED** [review]

**Comment by thatch45** [comment]
I am really sorry I missed this! Looks great to me!

File: salt/modules/state.py [hunk 1]
```diff
@@ -2133,6 +2133,34 @@ def pkg(pkg_path,
...
+        if not (s.startswith('debug_salt_state') and s.endswith('.tgz')):
+            continue
+        pkg_path = os.path.join(pkg_dir, s)
+        pkgs.append((pkg_path, os.stat(pkg_path).st_mtime))
+    n_to_keep = 10
+    for pkg_path, _ in sorted(pkgs, key=lambda pkg: pkg[1])[:-n_to_keep]:
+        os.unlink(pkg_path)
+
+
+def pkg_ssh(pkg_path, *args, **kwargs):
```
Comment (rallytime) [inline salt/modules/state.py]: Can you add some documentation to this new function?

Comment (mephi42) [reply salt/modules/state.py]: Ok, will do.

Comment (mephi42) [reply salt/modules/state.py]: I realized that I don't, strictly speaking, need a new function here.

So I introduced a new parameter remove_pkg for the existing pkg(), which is False by default. salt-ssh sets it to True, and this triggers the new behavior.

**Review by gtmanfred — APPROVED** [review]

**Comment by gtmanfred** [comment]
This has been confirmed to not break the `docker.call` stuff that uses this same wrapper state stuff right?

re-run py

**Comment by gtmanfred** [comment]
re-run py

File: salt/modules/state.py [hunk 2]
```diff
@@ -2133,6 +2125,46 @@ def pkg(pkg_path,
...
+        salt '*' state.pkg /tmp/salt_state.tgz 760a9353810e36f6d81416366fc426dc md5
+    '''
+    try:
+        return _pkg(pkg_path, *args, **kwargs)
+    finally:
+        if remove_pkg:
+            # Keep the copy for debugging purposes.
+            pkg_dir = os.path.dirname(pkg_path)
+            debug_pkg_path = os.path.join(
+                pkg_dir, 'debug_%s' % os.path.basename(pkg_path))
```
Comment (gtmanfred) [inline salt/modules/state.py]: please use format strings here.

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/22650/violations/file/salt/modules/state.py/

Or just `'debug_' + os.path.basename(pkg_path)` which is faster.

**Review by gtmanfred — CHANGES_REQUESTED** [review]
One lint failure, and wanting to make sure the docker tests that I added that use some of this stuff for docker.highstate and docker.sls don't fail.

**Comment by rallytime** [comment]
Hi @mephi42 - We haven't heard from you in a while. Are you able to make the changes as discussed above?

**Comment by mephi42** [comment]
Hi @rallytime, sorry, I'm a tad busy with other things at the moment, but this is still on my list of TODOs.

**Comment by rallytime** [comment]
No problem - I am going to close this for now. But please let us know when you're ready to come back to this and we will gladly re-open! Or you can open a new PR and link back to this one. Whichever is easier for you.

Do note however that there is a merge-conflict in one file presently. That will need to be fixed as well when you're ready to come back here.

The conflict is in `salt/client/ssh/wrapper/state.py`. Thanks!

**Pull request closed without merging**